src/agent/agent.cc

```diff
@@ -0,0 +1,490 @@
+#ifdef SUPPORT_AGENT_FORWARDING
+  if (comm == NULL) {
+    return;
+  }
+  AgentBuffers::Instruction inst;
+  inst.set_agent_id(agent_id);
+  string pb_packet;
+  fatal_assert(inst.SerializeToString(&pb_packet));
+  comm->send(pb_packet);
+#endif
+}
+
+
+AgentConnection *ProxyAgent::get_session() {
+#ifdef SUPPORT_AGENT_FORWARDING
+  if ( (! server) || l_sock < 0) {
+    return NULL;
+  }
+  struct sockaddr_un sunaddr; 
+  socklen_t slen = sizeof ( sunaddr );
+  memset( &sunaddr, 0, slen );
+  int sock = accept ( l_sock, (struct sockaddr *)&sunaddr, &slen );
+  if ( sock < 0 ) {
+    return NULL;
+  } 
+
+  if ( (comm == NULL) || (oob_ctl_ptr == NULL) ) {
+     (void) close( sock );
+     return NULL;
+  }
+
+  /* Here we should check that peer effective uid matches with the
+     euid of this process.  Skipping however and trusting the file
+     system to protect the socket. This would basically catch root
+     accessing the socket, but root can change its effective uid to
+     match the socket anyways, so it doesn't really help at all. */
+
+  /* If can't set the socket  mode, discard it. */
+  if ( fcntl( sock, F_SETFD, FD_CLOEXEC ) != 0 || fcntl( sock, F_SETFL, O_NONBLOCK ) != 0 ) {
+     (void) close( sock );
+     return NULL;
+  }
+  return new AgentConnection ( sock, ++cnt, this );
+#else
+  return NULL;
+#endif
+}
+
+AgentConnection::AgentConnection(int sock, uint64_t id, ProxyAgent *s_agent_ptr)
+  : s_in_read_set( false ),
+    s_sock( sock ),
+    s_id( id ),
+    idle_start( Network::timestamp() ),
+    packet_buf( "" ),
+    packet_len( 0 ),
+    agent_ptr( s_agent_ptr )
+{
+#ifndef SUPPORT_AGENT_FORWARDING
+  if (sock >= 0) {
+    (void) close( sock );
+  }
+  s_sock = -1;
+#endif
+}
+
+AgentConnection::AgentConnection(std::string agent_path, uint64_t id, ProxyAgent *s_agent_ptr)
+  : s_in_read_set( false ),
+    s_sock( -1 ),
+    s_id( id ),
+    idle_start( Network::timestamp() ),
+    packet_buf( "" ),
+    packet_len( 0 ),
+    agent_ptr( s_agent_ptr )
+{
+#ifdef SUPPORT_AGENT_FORWARDING
+  int sock = socket( AF_UNIX, SOCK_STREAM, 0 );
+  struct sockaddr_un sunaddr;
+  memset( &sunaddr, 0, sizeof (sunaddr) );
+  sunaddr.sun_family = AF_UNIX;
+  if ( agent_path.length() >= sizeof (sunaddr.sun_path) ) {
+    (void) close( sock );
+    return;
+  }
+  if ( fcntl( sock, F_SETFD, FD_CLOEXEC ) != 0 ) {
+    (void) close( sock );
+    return;
+  }
+  strncpy( sunaddr.sun_path, agent_path.c_str(), sizeof (sunaddr.sun_path) );
+  if ( connect(sock, (struct sockaddr *)&sunaddr, sizeof (sunaddr)) < 0 ) {
+    (void) close( sock );
+    return;
+  }
+  if ( fcntl( sock, F_SETFL, O_NONBLOCK ) != 0 ) {
+    (void) close( sock );
+    return;
+  }
+  s_sock = sock;
+#endif
+}
+
+AgentConnection::~AgentConnection() {
+  if ( s_sock >= 0 ) {
+    (void) close ( s_sock );
+  }
+}
+
+uint64_t AgentConnection::idle_time() {
+  return (Network::timestamp() - idle_start) / 1000;
+}
+
+string AgentConnection::recv_packet() {
```

Comment:
This method has an issue: it assumes the entire agent-forwarding message is immediately available.  If it were to encounter a client that does unbuffered i/o and does separate writes for the message length and message data, it would fail unpredictably.  This should be reworked to an event driven style.